gap/lqueue.gi

```diff
@@ -1,174 +1,203 @@
-#############################################################################
-##
-#W  lqueue.gd                   GAP library                  Reimer Behrends
-##
-##
-#Y  Copyright (C) 2013 The GAP Group
-##
-##  This file implements a deque based on a circular buffer. It can be used
-##  to implement FIFO queues as well as stacks.
-##
+# lqueue.gi
+
+# This file implements double ended queues (deques) using a circular buffer
+# stored in a GAP plain list.
+#
+# The four positions in a queue Q have the following purpose
+#
+# Q[1] - head, the index in Q[4] of the first element in the queue
+# Q[2] - tail, the index in Q[4] of the last element in the queue
+# Q[3] - capacity, the allocated capacity in the queue
+# Q[4] - GAP Plain list with storage for capacity many entries
+#
+# When the queue fills up the capacity is doubled using PlistQueueExpand,
+# This is currently done when after pushing to the front or the back of the
+# queue fills the last free place. A new plist is allocated and all current
+# entries of the queue are copied into the new Plist with the head entry being
+# at index 1 and the tail entry being at entry "old capacity".
+#
+# The queue is empty if and only if head = tail and the entry that head and tail
+# point to in the storage list is unbound.
 
 InstallGlobalFunction(PlistQueue,
 function(arg)
-  local capacity, filter, result, i, t;
-
-  filter := IsPlistQueueRep;
-
-  if Length(arg) = 0 then
-    capacity := 64;
-  elif Length(arg) = 1 then
-    capacity := arg[1];
-  fi;
+    local capacity, result;
 
-  result := [1, 1, capacity, EmptyPlist(capacity)];
-  for i in [1..capacity] do
-    result[4][i] := fail;
-  od;
-
-  t := NewType(CollectionsFamily(FamilyObj(IsObject)), filter and IsPositionalObjectRep);
+    if Length(arg) = 0 then
+        capacity := 64;
+    elif Length(arg) = 1 then
+        capacity := arg[1];
+    fi;
 
-  Objectify(t, result);
+    result := [1, 1, capacity, EmptyPlist(capacity)];
+    Objectify(PlistQueueType, result);
 
-  return result;
+    return result;
 end);
 
 InstallGlobalFunction(PlistQueuePushBack,
- function(queue,el)
-  local head, tail, last;
-  head := queue![QHEAD];
-  tail := queue![QTAIL];
-  last := queue![QCAPACITY];
-  if tail = last then
-    if head = 1 then
-      PlistQueueExpand(queue);
-      tail := queue![QTAIL];
-      queue![QDATA][tail] := el;
-      tail := tail + 1;
-      queue![QTAIL] := tail;
+function(queue,el)
+    local head, tail, last;
+    head := queue![QHEAD];
+    tail := queue![QTAIL];
+    last := queue![QCAPACITY];
+
+    # Special case for empty queue,
+    # head = tail and queue![QDATA][tail] is
+    # not bound
+    if not IsBound(queue![QDATA][tail]) then
+        queue![QDATA][tail] := el;
     else
-      queue![QTAIL] := 1;
-      queue![QDATA][last] := el;
+        if tail = last then
+            tail := 1;
+        else
+            tail := tail + 1;
+        fi;
```

Comment:
Note that there is a global `last`, which I find mildly confusing.
We could actually get rid of `QCAPACITY` with the following trick: Just bind `queue![QDATA][last+1]` to something, say `fail`. Then replace `last` by `Length(queue![QDATA])-1` or equivalent.

The above check could become `if tail >= Length(queue![QDATA])-1 then`